File: frontend/src/utils/stringValues.js
export function emptyIfNull(value) {
  return value === null || value === undefined ? '' : String(value);
}

export function trimToEmpty(value) {
  return emptyIfNull(value).trim();
}

export function normalizeWhitespace(value) {
  return trimToEmpty(value).replace(/\s+/g, ' ');
}

export function normalizeSearchKey(value) {
  return normalizeWhitespace(value).toLowerCase();
}

export function firstNonBlank(...values) {
  for (const value of values) {
    const normalized = trimToEmpty(value);
    if (normalized) return normalized;
  }
  return null;
}

export function containsAny(text, words = []) {
  const source = emptyIfNull(text);
  return words.some((word) => word !== null && word !== undefined && source.includes(String(word)));
}

export function uniqueNonBlank(values) {
  return [...new Set((values || [])
    .map((value) => trimToEmpty(value))
    .filter(Boolean))];
}

export function sortedByLengthDesc(values) {
  return [...(values || [])].sort((a, b) => String(b).length - String(a).length);
}

File: backend/src/main/java/com/comfortableledger/ledger/util/StringValues.java
package com.comfortableledger.ledger.util;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

public final class StringValues {
    private StringValues() {
    }

    public static String emptyIfNull(String value) {
        return value == null ? "" : value;
    }

    public static String trimToEmpty(String value) {
        return value == null ? "" : value.trim();
    }

    public static String normalizeWhitespace(String value) {
        return value == null ? "" : value.trim().replaceAll("\\s+", " ");
    }

    public static String normalizeSearchKey(String value) {
        return normalizeWhitespace(value).toLowerCase(Locale.ROOT);
    }

    public static String firstNonBlank(String... values) {
        if (values == null) {
            return null;
        }
        for (String value : values) {
            String normalized = trimToEmpty(value);
            if (!normalized.isBlank()) {
                return normalized;
            }
        }
        return null;
    }

    public static boolean containsAny(String text, String... words) {
        String source = emptyIfNull(text);
        if (words == null) {
            return false;
        }
        for (String word : words) {
            if (word != null && source.contains(word)) {
                return true;
            }
        }
        return false;
    }

    public static List<String> sortedByLengthDesc(Collection<String> words) {
        if (words == null) {
            return List.of();
        }
        return words.stream()
                .sorted(Comparator.comparingInt(String::length).reversed())
                .toList();
    }
}
